README.md
LGBinExtractor
==============

A tool for extracting bin firmware for LG Phones

http://forum.xda-developers.com/showthread.php?t=1879915

##NOTES

 1. Sometimes partition require that they fill the entire 'size' even though the last bit is whitespace. The bin/tot files often don't have that last bit of whitespace. The problem is only noticed so far in tot partitions that are broken up into seperate parts and there is a prompt to ask you if you want to add the data if there is a GPT header available to get correct size from. I intend to later analyze the partition itself to see if it's neccesary and how much 'space' should there be.

 2. My database of bin files is still to small to analyze them when it's unkown and only supported bin/tot files will work. Just follow the steps present in Support to get it added

##Contributing
Made a change? </br>
Fixed a bug?</br>
Made a improvement?</br>

Don't keep it to yourself. Make a pull request.

https://github.com/Xonar/LGBinExtractor/pull/new/master

###HALL OF HELPERS
 - novassoc : for his tool that extracts the kdz and wdb files.
 - SnowLeopardJB : for lots and lots of testing

##Bugs
Found a bug? Please report it!

https://github.com/Xonar/LGBinExtractor/issues

I'll do my best to fix it as quickly as possible.

###Known bugs
 - Doesn't work with Big Endian Systems
 - Some partitions require that they are their full size (see Note 1)

##Support
If there isn't support for your phone just ask. I'll be glad to include it. As I don't own many of these phones (Just the P970 is mine) and my bandwidth is limited I need some info from you if I am to add support.

######EMail or msg me on XDA with the following info:
- Phone Model
- What firmware you want to extract
- The first meg of the firmware file (Theres a option in the tool for that now)</br>
 - It is rare that I can actually download the very large firmware files so this is important</br>
 - Any hosting site will be sufficient. You can also post in the thread and attach the file there.
- Your phone partition info</br>
 - the output from 'cat /proc/partitions' is usually sufficient
- All known partition starting points within the file.

You should also be prepared for some correspondence so I can use you to test if it actually works. I am doing it for you after all.
    	
This tool is pretty picky when it comes to magic checking. I'll add support for forcing a certain magic and ignoring magic checking later. I have a life of my own after all.

###CONFIRMED WORKING FOR :
See Note 2 and Support

#####BIN:
 - P970
 - KU5900
 - AS730
 - P940
 - LG Lucid

#####TOT:
- Optimus G (Sprint,AT&T,Bell)
- Optimus G Pro
- Nexus 4
- MS770

###TESTING/PLANNED NEAR FUTURE
- P990

##CONTACT INFO
E-Mail : xonar.leroux@gmail.com</br>
XDA	: http://forum.xda-developers.com/member.php?u=4669225

If you want common info then please ask it on the thread and don't message me directly.
If you want to make a contribution just message me with what you want to do or make a pull request with what you've done.


CHANGELOG
CHANGELOG

03/05/2013 #2
	Add Support for LG L9

03/05/2013 #1
	Improved APHeader display on failed read

05/04/2013
	Fix reading 44 EC 55 AA header

01/04/2013 #2
	Don't assume strings are null terminated

01/04/2013 #1
	Add support for LG Lucid
	Fix Readme mistake
	Fix Makefile mistake

30/03/2013 #3
	Added support for Motion 4G

30/03/2013 #2
	Added support for Optimus G Pro

30/03/2013 #1
	Added DEBUG prints

23/01/2013
	Added Makefile

22/01/2013
	Fix adding 1 to little LBA of whitespace

21/01/2013 #3
	Fix adding way to much whitespace
	Update Readme

21/01/2013 #2
	Add whitespace after merged partitions.
	Align output after merged partitions

21/01/2013 #1
	Improve cmdline parsing
	Change writing space in merged partitions.

20/01/2013 #4
	Revamp reading 44 DD 55 AA APHeader
	Add Ability to extract header

20/01/2013 #3
	Add support for P940

20/01/2013 #2
	Improve HEX Output

20/01/2013 #1
	Fix: Not reading file in binary in some OS
	Fix: Initialized some values to NULL/0
	Improve Code Formatting

19/01/2013
	Fix Markup on Readme

18/01/2013 #3
	Fix: Segfault when opening file that can't be opened
	Improved Comments and Console Output

18/01/2013 #2
	Fix: Merge Blocks when Same Name

18/01/2013 #1
	Merge Blocks when Same Name
	Misc changes

16/01/2013 #2
	Add support for Sprint Optimus G

16/01/2013 #1
	Add Test support for the AT&T Optimus G.
	Improved Magic Checking.
	Change it a Bit to more Easily Support Mutliple Magic Numbers.

02/01/2013
	Change to C89 Standard for better cross-platform compatibilaty

18/12/2012
	Fix possible cross-platform incompatibilaty
	Small design changes to allow multiple bin formats
	Added support for new bin format : 44 DD 55 AA
	Better code formatting
	Added more error checks
	Fix not removing first arg correctly when binary is not correctly capatalized
	Added Changelog :p

20/11/2012 #2
	Add Commandline Interface
	Better code comments
	Slight code improvements and format changes
	Remove unnecesary inlcudes

20/11/2012 #2
	Write files in 512 blocks
	Updated ReadMe

19/11/2012
	Initial Release
	


src/GPT.c
/*
 * GPT.c
 *
 *  Created on: Nov 18, 2012
 *      Author: xonar
 */

#include "GPT.h"
#include "BinExtractor.h"

GPTHeader readGPTHeader(FILE* f)
{
    /*READ GPT HEADER INTO STRUCT*/
    GPTHeader out;
    fread(&out, sizeof(GPTHeader), 1, f);
    return out;
}

GPTPartitionEntry readGPTPartitionEntry(FILE* f)
{
    /*Declarations*/
    GPTPartitionEntry out;
    int i = 1, j = 2;

    fread(&out, sizeof(GPTPartitionEntry), 1, f);

    /*Do poor mans convention from UTF16 to UTF8
     Ignoring all non ASCII characters*/

    for(; i < 32; i++, j += 2)
        out.part_name[i] = out.part_name[j];

    out.part_name[32] = '\0';

    return out;
}

void readGPTPartitionEntryArray(FILE* f, GPTPartitionEntry* out, int num)
{
    fread(out, sizeof(GPTPartitionEntry), num, f);

    /*Do poor mans convention from UTF16 to UTF8
     Ignoring all non ASCII characters*/

    num--;
    for(; num >= 0; num--)
    {
        int i = 1;
        int j = 2;
        for(; i < 32; i++, j += 2)
            out[num].part_name[i] = out[num].part_name[j];

        out[num].part_name[32] = '\0';
    }
}

void printGPTHeader(const GPTHeader h, FILE* f)
{
    fprintf(f, "GPT HEADER\n----------\n%-30s", "Signature");

    printHex(f, h.signature, 8);

    fprintf(f, "\n%-30s%" PRIu32 "\n"
    "%-30s%" PRIu32 "\n"
    "%-30s", "Revision", h.revision, "Header Size", h.hsize, "CRC32 of Header");

    printHex(f, &h.crc,sizeof(uint32_t));

    fprintf(f, "\n%-30s%" PRIu64 "\n"
    "%-30s%" PRIu64 "\n"
    "%-30s%" PRIu64 "\n"
    "%-30s%" PRIu64 "\n"
    "%-30s", "Current Header LBA", h.current_lba, "Backup Header LBA", h.backup_lba,
            "First Usable LBA", h.first_lba, "Last Usable LBA", h.last_lba, "Disk GUID");

    printHex(f, h.disk_guid, 16);

    fprintf(f, "\n%-30s%" PRIu64 "\n"
    "%-30s%" PRIu32 "\n"
    "%-30s%" PRIu32 "\n"
    "%-30s", "Start of Partition Entries", h.pent_lba, "Number of Partition Entries", h.pent_num,
            "Size of Partition Entries", h.pent_size, "CRC32 of Partition Array");

    printHex(f, &h.crc_part,sizeof(uint32_t));

    fputc('\n', f);
}

void printGPTPartitionEntry(const GPTPartitionEntry pe, FILE* f)
{
    fprintf(f, "PARTITION ENTRY\n---------------\n    %-26s", "Partition Type GUID");

    printHex(f, pe.ptype_guid, 16);

    fprintf(f, "\n    %-26s", "Unique Partition GUID");

    printHex(f, pe.upart_guid, 16);

    fprintf(f, "\n    %-26s%" PRIu64 "\n"
    "    %-26s%" PRIu64 "\n"
    "    %-26s%" PRIu64 "\n"
    "    %-26s%s\n", "First LBA", pe.first_lba, "Last LBA", pe.last_lba, "Attributes", pe.att_flags,
            "Partition Name", pe.part_name);
}

void printFullGPTInfo(const GPTHeader h, const GPTPartitionEntry* pe, FILE* f)
{
    int i = 0;
    GPTPartitionEntry PE_NULL =
    {
    { 0 },
    { 0 }, 0, 0, 0, "" };

    printGPTHeader(h, f);

    fprintf(f, "\nPARTITION ENTRIES\n-----------------\n");

    for(; i < h.pent_num && memcmp(&pe[i], &PE_NULL, sizeof(GPTPartitionEntry)); i++)
    {
        fputs("\n", f);
        printGPTPartitionEntry(pe[i], f);
    }
}


src/GPT.h
#ifndef GPT_H
#define GPT_H

#include <inttypes.h>
#include <stdio.h>
#include <string.h>

/*GPT HEADER STRUCT*/
typedef struct _GPTHeader
{
    char signature[8];
    uint32_t revision;
    uint32_t hsize;
    uint32_t crc;
    uint32_t reserved;
    uint64_t current_lba;
    uint64_t backup_lba;
    uint64_t first_lba;
    uint64_t last_lba;
    unsigned char disk_guid[16];
    uint64_t pent_lba;
    uint32_t pent_num;
    uint32_t pent_size;
    uint32_t crc_part;
    char end[420];
} GPTHeader;

/*PARTITION ENTRY STRUCT*/
typedef struct _GPTPartitionEntry
{
    unsigned char ptype_guid[16];
    unsigned char upart_guid[16];
    uint64_t first_lba;
    uint64_t last_lba;
    uint64_t att_flags;
    char part_name[72];
} GPTPartitionEntry;

/*READ*/
GPTHeader readGPTHeader(FILE* f);
GPTPartitionEntry readGPTPartitionEntry(FILE* f);
void readGPTPartitionEntryArray(FILE* f, GPTPartitionEntry* out, int num);

/*DISPLAY*/
void printGPTHeader(const GPTHeader h, FILE* f);
void printGPTPartitionEntry(const GPTPartitionEntry pe, FILE* f);
void printFullGPTInfo(const GPTHeader h, const GPTPartitionEntry* pe, FILE* f);

#endif


src/APHeader.h
/*
 * APHeader.h
 *
 *  Created on: Nov 18, 2012
 *      Author: xonar
 */

#ifndef APHEADER_H
#define APHEADER_H

#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include "BinExtractor.h"

/*PARTITION ENTRY STRUCT*/
typedef struct _APPartitionEntry
{
    char name[257];/*256 + 1 for terminating char*/
    uint32_t pent_id;
    uint32_t file_off;
    uint32_t file_size;
    uint32_t disk_size;
    uint32_t disk_off;
} APPartitionEntry;

/*AP HEADER STRUCT*/

/*MAGIC NUMBER*/
typedef struct _MagicNumber
{
    uint32_t magic;
    unsigned int off;
    struct _MagicNumber* next;
} MagicNumber;

#define INIT_MAGICNUMBER(mn)                \
        mn.magic = 0;                       \
        mn.off = 0;                         \
        mn.next = NULL;

typedef struct _APHeader
{
    MagicNumber magic;
    int pent_num;
    APPartitionEntry* pent_arr;
} APHeader;

#define INIT_APHEADER(aph)                  \
   aph.pent_num = 0;                        \
   aph.pent_arr = NULL;                     \
   aph.magic.magic = 0;                     \
   aph.magic.off = 0;                       \
   aph.magic.next = NULL;

typedef enum _DataType
{
    DISK_SIZE, DISK_OFF, FILE_SIZE, FILE_OFF, BLOCK_ID, BLOCK_NAME, SKIP
} DataType;

typedef struct _Item
{
    DataType type;
    size_t size;
} Item;

/*DATA BLOCK*/
typedef struct _DataBlock
{
    size_t blockOff;
    size_t blockSize;
    uint8_t numItems;
    Item* items;
    struct _DataBlock* next;
} DataBlock;

#define INIT_DATABLOCK(db)                  \
        db.blockOff = 0;                    \
        db.blockSize = 0                    \
        db.numBlocks = 0;                   \
        db.numItems = 0;                    \
        db.items = NULL;                    \
        db.next = NULL;

/*READ*/
APHeader readAPHeader(FILE *f);

APHeader readAPHeader44DD55AA(FILE *f);
APHeader readAPHeader44EC55AA(FILE *f);
APHeader readAPHeaderA5A555AA(FILE *f);

/*DISPLAY*/
void printAPHeader(const APHeader h, FILE *f);
void printAPPartitionEntry(const APPartitionEntry pe, FILE *f);
void printFullAPInfo(const APHeader h, FILE *f);

#endif /* APHEADER_H_ */


src/APHeader.c
/*
 * APHeader.c
 *
 *  Created on: Nov 18, 2012
 *      Author: xonar
 */

#include "APHeader.h"

APHeader readAPHeader(FILE *f)
{
    APHeader out;
    uint32_t* magic = (uint32_t*) &out.magic;

    INIT_APHEADER(out);

    /*READ MAGIC NUMBER*/
    fread(&out, sizeof(char) * 4, 1, f);

    /*PASS TO CORRECT FUNCTION BASED ON MAGIC NUMBER*/
    switch (*magic)
    {
    case 0xaa55ec44:
        out = readAPHeader44EC55AA(f);
        break;
    case 0xaa55a5a5:
        out = readAPHeaderA5A555AA(f);
        break;
    case 0xaa55dd44:
        out = readAPHeader44DD55AA(f);
        break;
    default:
        fprintf(stderr, "Unsupported Bin Type : ");
        printHex(stderr, magic, sizeof(*magic));
        fprintf(stderr, "\n");
        break;
    }

    return out;
}

APHeader readAPHeader44EC55AA(FILE *f)
{
    APHeader out;
    int i = 0;

    INIT_APHEADER(out);

    out.magic.magic = 0xaa55ec44;

    /*COUNT AND ALLOCATE AP ENTRIES*/
    while (1)
    {
        uint64_t tmp;
        fread(&tmp, sizeof(tmp), 1, f);
        if (tmp == UINT64_MAX) break;
        else out.pent_num++;
    }

    fseek(f, 4, SEEK_SET);

    out.pent_arr = (APPartitionEntry*) calloc(out.pent_num,
            sizeof(APPartitionEntry));

    /*READ FILE OFFSET AND FILE SIZE*/
    for (; i < out.pent_num; i++)
    {
        fread(&out.pent_arr[i].file_off, sizeof(uint32_t), 1, f);
        out.pent_arr[i].disk_off = out.pent_arr[i].file_off;
        fread(&out.pent_arr[i].file_size, sizeof(uint32_t), 1, f);
    }

    /*READ ID, DISK SIZE AND NAME*/
    fseek(f, 0x200, SEEK_SET);

    for (i = 0; i < out.pent_num; i++)
    {
        fread(&out.pent_arr[i].pent_id, sizeof(uint32_t), 1, f);
        fread(&out.pent_arr[i].disk_size, sizeof(uint32_t), 1, f);
        fseek(f, 0x4, SEEK_CUR);
        fread(&out.pent_arr[i].name, sizeof(char), 0x14, f);
        out.pent_arr[i].name[0x14] = '\0';
        fseek(f, 0x1e0, SEEK_CUR);
    }

    return out;
}

APHeader readAPHeaderA5A555AA(FILE *f)
{
    APHeader out;
    int i = 0;

    INIT_APHEADER(out);

    _Bool alternate = 0;

    /*DETERMINE ALTERNATE A5A555AA*/
    {
        char tmp;
        fseek(f,0x20c,SEEK_SET);
        fread(&tmp, sizeof(tmp), 1, f);
        if ((tmp >= 'A' && tmp <= 'Z') || (tmp >= 'a' && tmp <= 'z')
                || (tmp >= '0' && tmp <= '9')) alternate = 1;
        fseek(f, 4, SEEK_SET);
    }

    out.magic.magic = 0xaa55a5a5;

    /*COUNT AND ALLOCATE AP ENTRIES*/
    while (1)
    {
        uint64_t tmp;
        fread(&tmp, sizeof(tmp), 1, f);
        if (tmp == UINT64_MAX) break;
        else out.pent_num++;
    }

    fseek(f, 4, SEEK_SET);

    out.pent_arr = (APPartitionEntry*) calloc(out.pent_num,
            sizeof(APPartitionEntry));

    /*READ FILE OFFSET AND SIZE*/
    for (; i < out.pent_num; i++)
    {
        fread(&out.pent_arr[i].file_off, sizeof(uint32_t), 1, f);
        fread(&out.pent_arr[i].file_size, sizeof(uint32_t), 1, f);
    }

    skipToNextLBA(f);

    /*READ DISK SIZE IGNORING FIRST ID REFERENCE*/
    for (i = 0; i < out.pent_num; i++)
    {
        if (!alternate)
        {
            /*fread(&out.pent_arr[i].pent_id,sizeof(uint32_t),1,f);*/
            fseek(f, 4, SEEK_CUR);
            fread(&out.pent_arr[i].disk_size, sizeof(uint32_t), 1, f);
            fseek(f, 504, SEEK_CUR);
        }
        /*ALTERNATIVE*/
        else
        {
            fread(&out.pent_arr[i].pent_id, sizeof(uint32_t), 1, f);
            fread(&out.pent_arr[i].disk_size, sizeof(uint32_t), 1, f);
            fseek(f, 4, SEEK_CUR);
            fread(&out.pent_arr[i].name, sizeof(char), 20, f);
            fseek(f, 480, SEEK_CUR);
            out.pent_arr[i].name[256] = '\0';
            out.pent_arr[i].disk_off = 0xFFFFFFFF;
        }
        /*END OF ALTERNATIVE*/
    }

    fseek(f, 0x2200, SEEK_SET);

    /*READ ID AND NAME AND SET DISK OFF*/
    if (!alternate) for (i = 0; i < out.pent_num; i++)
    {
        fread(&out.pent_arr[i].pent_id, sizeof(uint32_t), 1, f);
        fread(&out.pent_arr[i].name, sizeof(char), 256, f);
        out.pent_arr[i].name[256] = '\0';
        out.pent_arr[i].disk_off = 0xFFFFFFFF;
    }

    return out;
}

APHeader readAPHeader44DD55AA(FILE *f)
{
    APHeader out;
    int i = 4, j;
    char buf[512];

    DataBlock dataBlock, *curDataBlock;
    MagicNumber *curMagicNum;
    uint32_t tmp;

    INIT_APHEADER(out);

    /*READ FIRST MAGIC NUMER*/
    curMagicNum = &out.magic;

    fseek(f, 0, SEEK_SET);
    fread(&curMagicNum->magic, sizeof(uint32_t), 1, f);

    /*SCAN FOR MAGIC NUMBERS*/
    for (; i < 0x2004; i += 4)
    {
        fread(&tmp, sizeof(uint32_t), 1, f);

        if (tmp != 0xffffffff)
        {
            /*FOUND MAGIC NUM*/
            curMagicNum->next = malloc(sizeof(MagicNumber));
            curMagicNum = curMagicNum->next;
            curMagicNum->next = NULL;
            curMagicNum->magic = tmp;
            curMagicNum->off = i;
        }
    }

    /*START WITH 44 DD 55 AA and ENDING WITH 33 EC 55 AA*/
    if (out.magic.magic != 0xaa55dd44 || curMagicNum->magic != 0xaa55ec33)
    {
        fprintf(stderr, "Not 0x0    :44DD55AA\n    0x2000 :33EC55AA\n");
        goto failed;
    }

    curMagicNum = &out.magic;

    /*SET FIRST DATABLOCK*/
    dataBlock.blockOff = 0x2010;
    dataBlock.blockSize = 12;
    dataBlock.numItems = 4;

    dataBlock.items = calloc(sizeof(Item), 4);
    dataBlock.items[0].type = DISK_OFF;
    dataBlock.items[0].size = 4;
    dataBlock.items[1].type = FILE_OFF;
    dataBlock.items[1].size = 4;
    dataBlock.items[2].type = FILE_SIZE;
    dataBlock.items[2].size = 4;
    dataBlock.items[3].type = SKIP;
    dataBlock.items[3].size = 4;

    dataBlock.next = NULL;

    curDataBlock = &dataBlock;

    /*TOT FILES*/
    if (out.magic.next->off == 0x8 && out.magic.next->next->off == 0x2000)
        switch (out.magic.next->magic)
        {
        default:
            printf("Did not detect known magic numbers!\n"
                    "  The File Appears to be a TOT File.All known TOT Files\n"
                    "  have the same format with different magic numbers.\n"
                    "  Do you want to assume that and try? Y/N : ");

            scanf("%s", buf);

            if (!(buf[0] == 'Y' || buf[0] == 'y')) break;

        case 0x8978f62b:
        case 0x5062c8ea:
        case 0x49838b94:
        case 0xdebf33af:
        case 0x42ef4e39:
        case 0x0e65f034:
        case 0x95f57d8c:
            /*tot*/
            curDataBlock->next = malloc(sizeof(DataBlock));
            curDataBlock = curDataBlock->next;
            curDataBlock->blockOff = 0x4220;
            curDataBlock->blockSize = 0x20;

            curDataBlock->numItems = 1;
            curDataBlock->items = calloc(sizeof(Item), 1);
            curDataBlock->items[0].type = BLOCK_NAME;
            curDataBlock->items[0].size = 0x20;

            goto readBlocks;
        }

    if (out.magic.next->off == 0x600 && out.magic.next->next->off == 0x2000)
        switch (out.magic.next->magic)
        {
        case 0xcc00bbaa:
            curDataBlock->next = malloc(sizeof(DataBlock));
            curDataBlock = curDataBlock->next;
            curDataBlock->blockOff = 0x2400;
            curDataBlock->blockSize = 0x20;

            curDataBlock->numItems = 5;
            curDataBlock->items = calloc(sizeof(Item), 5);
            curDataBlock->items[0].type = BLOCK_ID;
            curDataBlock->items[0].size = 4;
            curDataBlock->items[1].type = DISK_SIZE;
            curDataBlock->items[1].size = 4;
            curDataBlock->items[2].type = SKIP;
            curDataBlock->items[2].size = 4;
            curDataBlock->items[3].type = BLOCK_NAME;
            curDataBlock->items[3].size = 0x14;
            curDataBlock->items[4].type = SKIP;
            curDataBlock->items[4].size = 0x1E0;

            goto readBlocks;
        }

    if (out.magic.next->off == 0x2000)
    {
        printf("%x %x\n", out.magic.next->magic, out.magic.next->off);
        /*CHECK FOR THIRD MAGIC NUM AT 0x4000*/
        fseek(f, 0x4000, SEEK_SET);
        fread(&tmp, sizeof(uint32_t), 1, f);

        if (tmp == 0x00000000)
        {
            curDataBlock->next = malloc(sizeof(DataBlock));
            curDataBlock = curDataBlock->next;
            curDataBlock->blockOff = 0x3004;
            curDataBlock->blockSize = 0x14;

            curDataBlock->numItems = 1;
            curDataBlock->items = calloc(sizeof(Item), 1);
            curDataBlock->items[0].type = BLOCK_NAME;
            curDataBlock->items[0].size = 0x14;
        }
        else if (tmp != 0xffffffff)
        {
            /*FOUND MAGIC NUM*/
            curMagicNum = out.magic.next;
            curMagicNum->next = malloc(sizeof(MagicNumber));
            curMagicNum = curMagicNum->next;
            curMagicNum->magic = tmp;
            curMagicNum->off = 0x4000;

            switch (curMagicNum->magic)
            {
            case 0x56781234:
                curDataBlock->next = malloc(sizeof(DataBlock));
                curDataBlock = curDataBlock->next;
                curDataBlock->blockOff = 0x4220;
                curDataBlock->blockSize = 0x20;

                curDataBlock->numItems = 1;
                curDataBlock->items = calloc(sizeof(Item), 1);
                curDataBlock->items[0].type = BLOCK_NAME;
                curDataBlock->items[0].size = 0x20;
                break;
            default:
                goto failed;
            }
        }
        else
        {
            curDataBlock->next = malloc(sizeof(DataBlock));
            curDataBlock = curDataBlock->next;
            curDataBlock->blockOff = 0x2404;
            curDataBlock->blockSize = 0x14;

            curDataBlock->numItems = 1;
            curDataBlock->items = calloc(sizeof(Item), 1);
            curDataBlock->items[0].type = BLOCK_NAME;
            curDataBlock->items[0].size = 0x14;
        }

        goto readBlocks;
    }

    /*SEARCH FOR NAMES*/
    goto failed;

    readBlocks:

    /*COUNT AND ALLOCATE AP ENTRIES*/
    fseek(f, 0x2010, SEEK_SET);

    fread(buf, sizeof(char), 0x10, f);

    while (*((uint32_t*) buf) != 0xFFFFFFFF)
    {
        out.pent_num++;
        fread(buf, sizeof(uint32_t), 4, f);
    }

    out.pent_arr = (APPartitionEntry*) calloc(out.pent_num,
            sizeof(APPartitionEntry));

    /*INIT AP PARTITION ENTRIES*/
    for (i = 0; i < out.pent_num; i++)
    {
        out.pent_arr[i].disk_off = 0xffffffff;
        out.pent_arr[i].file_off = 0xffffffff;
        out.pent_arr[i].disk_size = 0xffffffff;
        out.pent_arr[i].file_size = 0xffffffff;
        out.pent_arr[i].pent_id = 0xffffffff;
        out.pent_arr[i].name[0] = '\0';
    }

    /*READ APDATA*/
    curDataBlock = &dataBlock;

    while (curDataBlock != NULL)
    {
        fseek(f, curDataBlock->blockOff, SEEK_SET);

        for (i = 0; i < out.pent_num; i++)
        {
            for (j = 0; j < curDataBlock->numItems; j++)
            {
                switch (curDataBlock->items[j].type)
                {
                case DISK_SIZE:
                    fread(&out.pent_arr[i].disk_size,
                            curDataBlock->items[j].size, 1, f);
                    break;
                case DISK_OFF:
                    fread(&out.pent_arr[i].disk_off,
                            curDataBlock->items[j].size, 1, f);
                    break;
                case FILE_SIZE:
                    fread(&out.pent_arr[i].file_size,
                            curDataBlock->items[j].size, 1, f);
                    break;
                case FILE_OFF:
                    fread(&out.pent_arr[i].file_off,
                            curDataBlock->items[j].size, 1, f);
                    break;
                case BLOCK_ID:
                    fread(&out.pent_arr[i].pent_id, curDataBlock->items[j].size,
                            1, f);
                    break;
                case BLOCK_NAME:
                    fread(&out.pent_arr[i].name, sizeof(char),
                            curDataBlock->items[j].size, f);
                    out.pent_arr[i].name[curDataBlock->items[j].size] = '\0';
                    break;
                case SKIP:
                    fseek(f, curDataBlock->items[j].size, SEEK_CUR);
                    break;
                }
            }
        }

        curDataBlock = curDataBlock->next;
    }

    /*MAKE SURE ID IS SET*/
    for (i = 0; i < out.pent_num; i++)
        if (out.pent_arr[i].pent_id == 0xffffffff) out.pent_arr[i].pent_id = i;

    return out;
    failed: fprintf(stderr, "Failed Reading APHeader!\n");
    return out;
}

void printAPHeader(const APHeader h, FILE *f)
{
    /*PRINT MAGIC NUMBER(S)*/
    MagicNumber* tmp = h.magic.next;
    fprintf(f, "AP HEADER\n----------\n%-30s0x%-8X: ", "Magic Number",
            h.magic.off);

    printHex(f, &h.magic.magic, 4);

    while (tmp)
    {
        fprintf(f, "\n%-30s0x%-8X: ", "", tmp->off);
        printHex(f, &tmp->magic, 4);
        tmp = tmp->next;
    }

    /*NUMBER OF PARTITIONS*/
    fprintf(f, "\n%-30s%d\n", "Number of Partitions", h.pent_num);
}

void printAPPartitionEntry(const APPartitionEntry pe, FILE *f)
{
    fprintf(f, "PARTITION ENTRY\n------------\n"
            "    %-26s%s\n"
            "    %-26s%d\n"
            "    %-26s%" PRIu32 "\n",
            "Data Block Name", pe.name, "Data Block ID", pe.pent_id,
            "Size on File", pe.file_size);

    if (pe.file_off != 0xFFFFFFFF)
    {
fprintf    (f, "    %-26s%" PRIu32 "\n", "File Offset", pe.file_off);
}

    if (pe.disk_size != 0xFFFFFFFF)
    {
fprintf    (f, "    %-26s%" PRIu32 "\n", "Size on Disk", pe.disk_size);
}

    if (pe.disk_off != 0xFFFFFFFF)
    {
fprintf    (f, "    %-26s%" PRIu32 "\n", "Disk Offset", pe.disk_off);
}
}

void printFullAPInfo(const APHeader h, FILE *f)
{
    int i = 0;

    printAPHeader(h, f);

    fprintf(f, "\nPARTITION ENTRIES\n-----------------\n");

    for (; i < h.pent_num; i++)
    {
        fputs("\n", f);
        printAPPartitionEntry(h.pent_arr[i], f);
    }
}
